web-ui/src/test/java/com/tutrit/webclient/controller/CustomerControllerTest.java

```diff
@@ -0,0 +1,119 @@
+import org.springframework.test.web.servlet.MvcResult;
+import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;
+import org.springframework.test.web.servlet.result.MockMvcResultHandlers;
+import org.springframework.test.web.servlet.result.MockMvcResultMatchers;
+
+import java.util.Objects;
+import java.util.Optional;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.mockito.Mockito.when;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.model;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.view;
+
+@AutoConfigureMockMvc
+@SpringBootTest(classes = SpringContext.SpringConfig.class)
+class CustomerControllerTest {
+    @MockBean
+    CustomerGateway customerGateway;
+    @Autowired
+    MockMvc mockMvc;
+    @Autowired
+    ObjectMapper objectMapper;
+
+    @Test
+    void findCustomerById() throws Exception {
+        when(customerGateway.findCustomerById("234")).thenReturn(Optional.of((createCustomer())));
+
+        final MvcResult result = mockMvc
+                .perform(MockMvcRequestBuilders.get("/customers/234"))
+                .andDo(MockMvcResultHandlers.print())
+                .andExpect(MockMvcResultMatchers.status().isOk())
+                .andExpect(model().attribute("customer", Matchers.equalTo(createCustomer())))
+                .andExpect(view().name("customer-form.html"))
+                .andReturn();
+        Customer actualCustomer = (Customer) Objects.requireNonNull(result.getModelAndView()).getModel().get("customer");
+        assertEquals(createCustomer(), actualCustomer);
+    }
+
+    @Test
+    void findCustomerByIdIfCustomerNull() throws Exception {
+        when(customerGateway.findCustomerById("231")).thenReturn((customerIsNull()));
+
+        final MvcResult result = mockMvc
+                .perform(MockMvcRequestBuilders.get("/customers/231"))
+                .andDo(MockMvcResultHandlers.print())
+                .andExpect(MockMvcResultMatchers.status().isOk())
+                .andExpect(model().attribute("customer", Matchers.equalTo(expectedCustomer())))
+                .andExpect(view().name("customer-form.html"))
+                .andExpect(model().attribute("error", Matchers.equalTo("Customer not found")))
+                .andReturn();
+        Customer actualCustomer = (Customer) Objects.requireNonNull(result.getModelAndView()).getModel().get("customer");
+        assertEquals(expectedCustomer(), actualCustomer);
+
+
+    }
+
+    @Test
+    void saveCustomer() throws Exception {
+        final MvcResult mvcResult = mockMvc
+                .perform(MockMvcRequestBuilders.post("/customers/23?" +
+                                "customerId=32" +
+                                "&name=Vlad" +
+                                "&city=Brest" +
+                                "&phoneNumber=+3754478937721" +
+                                "&email=VladVrest@mail.com")
```

Comment:
this also could be in .param(String, String)

Reply:
I did